### DISCUSSION TITLE
Excitation of port 2 (coaxial) - unexpected results

### DISCUSSION THREAD

**Discussion by toammann** [thliebig/openEMS-Project#43] [Q&A Usage] [answered]
Hello,

first of all tank you for providing this incredable software! At first I could not belive something like this was availble in open source :),

I am trying to simulate a Coaxial (SMA) to Microstip RF Transition. The the following steps were taken to build the model:

1) Create a parametric model of the transistion and the Connector (approximate) in FreeCAD
2) Export the model as *.stl files
3) Export the parameter values as a textfile
4) Import the parameter values and 3D model files in a openEMS script
5) Create parametric meshing based on the FreeCAD model parameters

![model](https://user-images.githubusercontent.com/9382233/223577528-e75a34fd-0ccc-489b-9ba2-e3ffda71bd03.png)

At first I was quite happy with the results by exciting port 1 (Microstrip). The model could be tuned to to achieve decent matching.

![port1_excitation](https://user-images.githubusercontent.com/9382233/223574757-5d8a5cf9-227f-471c-9d0b-54471888cbca.jpg)

Now (to finalize the model) I tried to generate all four S-paramter values for a touchstone file export. This is the point where I am struggeling. The exciation of port 2 (the coaxial one) produces completely different s-parameter results than the port 1 excitation.

![port2_excitation](https://user-images.githubusercontent.com/9382233/223577490-af8a65a5-d174-46f8-a6f8-98080b373e66.jpg)

The scatting parameters of this model should be at least approximately symetrical / reciprocal. 

I ckecked the calculated port impecances. They look fine (around 50R). The meshlines were placed at the outer most coordinates of the coaxial outer and inner conductor.

![mesh](https://user-images.githubusercontent.com/9382233/223577547-36f950bd-caa7-4f5e-8603-98116521bc59.PNG)

Some Questions:
1)  Are there certain rules how to set the feed shift? 
2) Is there a minmum port length that must be specified?
3) Can you recommend some literature on the port topic? 
3) Is it possible to change which port should be excited inside of the CSX object? How can excite multiple ports sequentially without re-reunning the entire script?

Any suggestions how to narrow this down any further? I can share the model and the openEMS script. However it is rather lengthy and runs for about 20min.

Thank you in advance![](url)

**Comment by toammann** [comment, marked as answer]
Hello,
the problem may be related to the PLM boundary at the microstrip. I was focused on the exciation of port 2 for quite some time. However it seems that the PML at the Microstrip port fails for some reason and provides a bad match.

Here is a simulation result unsing a "MUR" Boundary at port 1 (mircostrip) while exciting port 2  (coaxial). This result is much better!

![port2_excitation_mur_at_xmax](https://user-images.githubusercontent.com/9382233/223587025-31ade2c3-bdbe-4a33-a963-fa00617271d8.jpg)

What is the deal here with the PML at the microstrip port?

![microstrip_port](https://user-images.githubusercontent.com/9382233/223587183-ff4cf81d-399e-4553-934b-6ce69f5b1662.PNG)

**Reply by thliebig** [reply]
Remember that the PML is 8 cells thick. Make sure it does not touch any port internals like the excitation or voltage and current probes... That goes into the feed shift topic. Make sure it is outside the pml.

As for one of your questions. You always have to run simulation per active port. There is no way around that.

**Reply by toammann** [reply]
Hello,

it is really strange, to my understanding there requirements are fullfilled. The probes at the micrsotrip port are placed 17 meshcells from the boudary.

![grafik](https://user-images.githubusercontent.com/9382233/223668377-631d7581-1f79-4009-a2f9-1bf3f9bf80ef.png)

On the coaxial side hte excitation is seperated 11 meshcells from the PML.

![grafik](https://user-images.githubusercontent.com/9382233/223669083-37d20748-5745-4998-8ce9-1ad5ab2af88e.png)

Simulation scipt and model:
[aisler_sma.zip](https://github.com/thliebig/openEMS-Project/files/10919016/aisler_sma.zip)

**Reply by thliebig** [reply]
Well it is S22 or port 2 that has this bad matching. Thus it is the coaxial port that is not working as it should? Right?

**Reply by thliebig** [reply]
Closely looking at it, it seems that large parts of the conductors are missing?
![Paraview looking at the output with "--debug-PEC"](https://user-images.githubusercontent.com/1634387/223805263-8635f5ca-306f-4528-a35d-a83543b8e19b.png)
I will have a closer look.

**Reply by thliebig** [reply]
A sorry, false alarm... You defined it as copper, not PEC... I will change that and have another look...

**Reply by thliebig** [reply]
I think I found it....
There is a meshing issue for copper "Copper_L01". You have a correct z-mesh on top and bottom, BUT, since this is an STL the upper and lower faces are not hit due to some floating point issue.
So you were right about the float issue, but I thought you meant the FDTD engine using floats ;)
Sorry for the misunderstanding...
![debug-PEC](https://user-images.githubusercontent.com/1634387/223807742-aa67c917-a138-4d2c-b89d-e5aabfb8df33.png)

**Reply by thliebig** [reply]
What is still puzzling is that for "copper material" (not PEC) this should not matter as much... I really see no obvious issue...

**Reply by thliebig** [reply]
Moving the z-line at 0 to -0.01 gives reasonable results with setting everything metal to PEC...
But you should check your z-mesh and undersize it for the metal layers a little bit I guess...

I think using PEC instead of copper should be ok. For this frequencies metal losses will be low in comparison to dielectric losses...

**Reply by toammann** [reply]
Hello,
well, if the problem is related to the meshing why does the simulation work when exciting the other port?

What tool do you use to visualize the meshed geometry? This looks like a very helpful debug tool.

I took a look at the fields in the XZ plane. The wave propagates as expected from port 2 towards port 1. Again, near the PML boundary at x_max everything looks strange. It seems that suddenly there is a significant field magnitude inside of the copper. I can also see the wave beeing reflected at the boundary (what is consistent to the s-parameters).

![fields_reflection_pml](https://user-images.githubusercontent.com/9382233/223830173-058a900f-9c91-43ab-a2ad-4a534b21f997.png)

I also added a pec solder joint at the bottom of the conductor to prevent radiation. I forgot to add this part. This speeds up the simulation as the engery decays much quicker.

**Reply by thliebig** [reply]
Oh doh, I just realize... PML is of course a "lossy material" so it only can "continue" dielectric properties and looses all "normal" conductivity.
Thus you cannot use the PML to continue metal with finite conductivity... You have to switch using PEC for everything going into the PML!

**Reply by thliebig** [reply]
But using PEC has the added problem that then the floating point accuracy is a problem because not everywhere the metal stays in one piece...

I use this to debug the PEC:
`RunOpenEMS( Sim_Path, Sim_CSX ,'--debug-PEC');`

Then a "PEC_dump.vtp" file is created you can review in Paraview. But it puts all PEC material in one view so it can get a bit messy quickly... Every line represents a PEC (grid) edge...

**Reply by thliebig** [reply]
> Thus you cannot use the PML to continue metal with finite conductivity... You have to switch using PEC for everything going into the PML!

I really have to have a look at that, because I could just treat high kappa as PEC and make it metal even inside the PML I guess... But not today or any time too soon..

**Reply by thliebig** [reply]
Well I lied, it was easy enough at least for a simple fix.
I have modified the PML to recognize every "Kappa" above 1000 S/m as a conductor and disable the PML for such cells...
And that seems to produce good results for your original setup... It also confirms that the PML with finite conductors were the problem...
But to update the windows builds will have to wait...

**Reply by toammann** [reply]
Wow, this sounds awesome! Thank you very much for your effort. 

I will Look into it in the evening.

**Reply by toammann** [reply]
Hello,
I  found some time to confirm also that the Copper / (non PEC) was the problem. To verify this I added addional bodies in the freeCAD model with the length of the MSL port. The MSL port line is PEC anyway to it is just a logical extension to make the ground also PEC.

![grafik](https://user-images.githubusercontent.com/9382233/224099556-c5121278-8caa-497e-a922-e56d5b0620a9.png)

Of course having the improved PML boundary is the better solution for the future. For now it works this way. Using PEC for everything is also a solution.

![e-field_pmc_gnd_plane](https://user-images.githubusercontent.com/9382233/224099826-be59e6c3-5890-4a6a-ab4d-330eb71871c6.png)

No more reflections and and leakage at the x_max PML! The scattering parameters are still not perfectly symmetrical but i call it close enough for now. This may be related to the GND-planes on the top layer. They are seperated by "only" 3*substrate_height. So there may be a slight influence on the line impedance which the MSL port can not catch.

 I look forward to the measurement results of this structure!

![port1](https://user-images.githubusercontent.com/9382233/224117271-a781724a-203b-4dda-95a9-82d63be6c786.jpg)
![port2](https://user-images.githubusercontent.com/9382233/224117283-1fa83ca2-9e47-46a3-8b66-7b9517a50f44.jpg)

--------------------------------------------------

However, there are two more points I could not solve. I do not know if they are worth investigating. I will mention them for the sake of completeness.

**1) Cylindrical STL imported bodies are not recoginzed by openEMS.**
I can not explain this. FreeCAD generates *stl. files for the inner conductor of the connector as well as for the Teflon substrate. There are plenty of meshlines across these bodies. 

When running the simulation I get the following warnings:
```
Warning: Unused primitive (type: PolyhedronReader) detected in property: PEC!
Warning: Unused primitive (type: PolyhedronReader) detected in property: Teflon!
```
They do not show up in the PEC-dump. As a workaround these primitives were modelled in openEMS directly by using AddCylinder() with identical measures. This works without a problem. The issue can be reproduced by adjusting the comment marks in line 316 to 326.

**2) How to excite two ports sequentially (one after another)**

To generate a touchstone file I would like to sequentially excite the two ports - one by one - with one script execution. The port defintion is stored in the CSX structure. Is it possible to modify which port is active after the port has been defined by AddPort()?

I am thinking of something like this:
```
...
generate model and mesh
...

// Create port 1, set port 1 active 
// Create port 2, set port 2 passive

WriteOpenEMS( [Sim_Path1 '/' Sim_CSX], FDTD, CSX );
RunOpenEMS( Sim_Path1, Sim_CSX ,'--debug-PEC');
f = linspace( f_start, f_stop, 1601 );
port = calcPort( port, Sim_Path1, f, 'RefImpedance', 50);
s11 = port{1}.uf.ref./ port{1}.uf.inc;
s21 = port{2}.uf.ref./ port{1}.uf.inc;

// Modify CSX  structure to set port 1 passive and port 2 active

WriteOpenEMS( [Sim_Path2 '/' Sim_CSX], FDTD, CSX );
RunOpenEMS( Sim_Path2, Sim_CSX ,'--debug-PEC');

port = calcPort( port, Sim_Path2, f, 'RefImpedance', 50);
s12 = port{1}.uf.ref./ port{2}.uf.inc;
s22 = port{2}.uf.ref./ port{2}.uf.inc;

//  write the touchstone file
```

Regards!

**Reply by toammann** [reply]
I was able to solve the last two remaining problems.

**1) Cylindrical STL imported bodies are not recoginzed by openEMS.**
The issue was related to the export in FreeCAD. So no blame on openEMS here! When exporting the cylindrical shapes as *.stl file a mesh is generated. The process performs an arc approximation of the cylinder surface.The curvature is broken down to straight line sections which effectively remove portions of the circle. Therefore, even though the mesh lines are placed at the correct radius/diameter they may not hit the cylinder boundary which yields into the above warnings (AppCSXCAD shows the *.stl mesh without further simplification, this is very useful if you are aware of this fact...).

I now use the Mesh workbench in FreeCAD to export the *.stl files. The workbench allows control of the arc approximation. In a python script this can be achieved by the following lines using the standard Mesher:
```python
mesh_ld = 1 # mm
mesh_ad = 20*math.pi/180  # 20°

label = "conn_outer_body"
body.append(App.ActiveDocument.getObjectsByLabel(label)[0])
Mesh.export(body, "./" + label + ".stl")

label = "conn_inner_cond"
shape =  Part.getShape(App.ActiveDocument.getObjectsByLabel(label)[0])
mesh=MeshPart.meshFromShape(Shape=shape, LinearDeflection=mesh_ld, AngularDeflection=mesh_ad, Relative=True)
mesh.removeDuplicatedPoints()
mesh.write("./" + label + ".stl")
```

**2) How to excite two ports sequentially (one after another)**
It actually works like described above.  It is possible to create two separate simulation folders for port1 and port2.  The *.xml file for the first excitation can be placed in the folder of port1. Then the excitation section of the CSX structure can be changed ``CSX.Properties.Excitation`` to excite port two. Finally a second *.xml file is generated from the modified CSX into folder two.

I will mark this as solved now!

**Reply by 0xCoto** [reply]
> When exporting the cylindrical shapes as *.stl file a mesh is generated. The process performs an arc approximation of the cylinder surface.The curvature is broken down to straight line sections which effectively remove portions of the circle. Therefore, even though the mesh lines are placed at the correct radius/diameter they may not hit the cylinder boundary which yields into the above warnings (AppCSXCAD shows the *.stl mesh without further simplification, this is very useful if you are aware of this fact...).

I do it this way with FreeCAD - think I'd run into the problem you mentioned, or does this look OK?
```py
# Correct relative origin
obj = FreeCAD.getDocument("Unnamed").getObject(str(object.Name))
__shape = Part.getShape(FreeCAD.getDocument("Unnamed").getObject(parent),part_name+str(object.Name)+".",needSubElement=False,refine=False)
FreeCAD.ActiveDocument.addObject("Part::Feature",str(object.Name)+"_transformed").Shape=__shape
obj = FreeCAD.getDocument("Unnamed").getObject(str(object.Name)+"_transformed")

# Export STL
Mesh.export([obj], dir + str(object.Label) + ".stl")
```

**Reply by toammann** [reply]
Hello,
using `Mesh.export()` was actually also my first attempt to export which caused the "unused primitive" warning. 

There is a [short article (Export_to_STL_or_OBJ)](https://github.com/FreeCAD/FreeCAD-documentation/blob/main/wiki/Export_to_STL_or_OBJ.md) in the FreeCAD documentation which describes the two available options to export a *.stl. When using `Mesh.export()` you end up in Method 1 mentioned in the documentation where you have no control about the arc approximation. 

In addition the results are dependent on the [FreeCAD import/export settings](https://wiki.freecad.org/Import_Export_Preferences/en) of the particular FreeCAD instance. So two installations with different settings will produce variying results which is - in my opinion - undesired. I use therefore the mesh workbench from now on.

The pyton FreeCAD script lines in my above reply use the mesh workbench directly. Several meshers are available in the Mesh workbench using  ["Create Mesh from Shape"](https://wiki.freecad.org/Mesh_FromPartShape).

----------------------------------------


**Visualization of the problem**:

The grey shapes are the *.stl exported mesh files. Parts in yellow color are the ideal representations of the shapes.

This is the standard setting of the "displayed angular deflection" which is used when you call `Mesh.export()`:

![part design shape view anglular deflection](https://user-images.githubusercontent.com/9382233/224537717-3136e75b-9978-4173-9886-7e03a7f78846.png)


What we want for proper meshing is that the *.stl meh has one point on each edge of the cylinder bounding box. With a very coarse max. angular deflection of 90° this is rather obvious:

![grafik](https://user-images.githubusercontent.com/9382233/224539370-e205f337-7573-4400-8f30-782a33284ff5.png)


Now let´s have a look at a result of the standard *.stl export with the 28.5° max. angular deflection setting:

![28.5deg close up](https://user-images.githubusercontent.com/9382233/224538507-790a2e9f-f3df-4667-9346-2837bb063a88.png)


On the right and left side the circle get´s cropped because the outer most point is not located on the bounding box. The mesh I specified at diameter/2 can never hit the circle because the *.stl export has cropped the it!

To fix this one have to choose an angular deflection that produces a suitable number of segments and ensures points on the bounding box at every 4 edges. E.g. for an max. angular deflection setting of 20°:

![angluar def. comp](https://user-images.githubusercontent.com/9382233/224539082-5666c39e-301f-4de7-9f8c-5053025aaf2c.png)


This can be achieved by either changing the  "Part Design Shape view" setting (Method 1) or using the preferred Mesh workbench (Method 2, see python lines above).

So the short answer is: It depends on your "displayed angular deflection"  FreeCAD preference settings.

Regards!

**Reply by 0xCoto** [reply]
Thanks for the detailed explanation!!

**Reply by toammann** [reply]
Hello Thorsten,
i have experimented with the "floating point / meshing" issue you described above. 

In order to have the edges in the z-dimension correctly recognized by the PEC-dumps I have to undersize the mesh line by about 100e-9 units. This value is unexpectedly large. Ok to be fair  I have no idea how big such floating point deviations can become. However there was a slight difference in the simulation result.

Can you explain a little what happens during the meshing in this case? I would also be interested why this "should not matter so much" for non-PEC materials as mentioned this above. 

Is the under sizing necessary in all three dimensions (x, y, z)? I suppose that this should be done in every case when using a *.stl import? This can become cumbersomepretty quickly and impossible in some cases where two metal parts are located on top on each other.

Thank you very much!
